src/lexer.rs

```diff
@@ -0,0 +1,320 @@
+
+    #[token("}")]
+    CloseBrace,
+
+    #[token("[")]
+    OpenBracket,
+
+    #[token("]")]
+    CloseBracket,
+
+    #[token(".")]
+    Chain,
+
+    #[token("..")]
+    Spread,
+
+    #[token("false", |_| false)]
+    #[token("true", |_| true)]
+    Boolean(bool),
+
+    #[regex(r"-?0x[0-9a-fA-F]+(_[0-9a-fA-F]+)*", parse_hexadecimal)]
+    #[regex(r"-?0o[0-7]+(_[0-7]+)*", parse_octal)]
+    #[regex(r"-?0b[01]+(_[01]+)*", parse_binary)]
+    #[regex(r"-[0-9]+(_[0-9]+)*", parse_negative_decimal)]
+    #[regex(r"[0-9]+(_[0-9]+)*", parse_positive_decimal)]
+    Integer(Integer),
+
+    #[regex(r"-?[0-9]+\.[0-9]*([eE][+-]?[0-9]+)?", |lex| lex.slice().parse::<f64>())]
+    Float(f64),
+
+    #[regex(r"\$[a-zA-Z_][a-zA-Z0-9_]*", |lex| lex.slice().trim_start_matches('$'))]
+    InputName(&'input str),
+
+    #[token("\"", parse_literal)]
+    Literal(Vec<StringPart<'input>>),
+
+    #[regex(r#"'(?:[^'\\]|\\.)*'|[^\s.=0-9\[\]{}"'-][^\s.=\[\]{}"']*"#, |lex| lex.slice().trim_matches('\''))]
+    Key(&'input str),
+}
+
+fn parse_positive_decimal<'input>(
+    lex: &mut logos::Lexer<'input, Token<'input>>,
+) -> Result<Integer, ParseIntError> {
+    lex.slice()
+        .replace("_", "")
+        .parse::<u64>()
+        .map(Integer::from)
+}
+
+fn parse_negative_decimal<'input>(
+    lex: &mut logos::Lexer<'input, Token<'input>>,
+) -> Result<Integer, ParseIntError> {
+    lex.slice()
+        .replace("_", "")
+        .parse::<i64>()
+        .map(Integer::from)
+}
+
+fn parse_hexadecimal<'input>(
+    lex: &mut logos::Lexer<'input, Token<'input>>,
+) -> Result<Integer, ParseIntError> {
+    let input = lex.slice().replace("_", "");
+    let is_negative = input.starts_with('-');
+    let hex_part = if is_negative {
+        &input[3..]
+    } else {
+        &input[2..]
+    };
+
+    if is_negative {
+        i64::from_str_radix(hex_part, 16).map(|n| Integer::from(-n))
+    } else {
+        u64::from_str_radix(hex_part, 16).map(Integer::from)
+    }
+}
+
+fn parse_octal<'input>(
+    lex: &mut logos::Lexer<'input, Token<'input>>,
+) -> Result<Integer, ParseIntError> {
+    let input = lex.slice().replace("_", "");
+    let is_negative = input.starts_with('-');
+    let octal_part = if is_negative {
+        &input[3..]
+    } else {
+        &input[2..]
+    };
+
+    if is_negative {
+        i64::from_str_radix(octal_part, 8).map(|n| Integer::from(-n))
+    } else {
+        u64::from_str_radix(octal_part, 8).map(Integer::from)
+    }
+}
+
+fn parse_binary<'input>(
+    lex: &mut logos::Lexer<'input, Token<'input>>,
+) -> Result<Integer, ParseIntError> {
+    let input = lex.slice().replace("_", "");
+    let is_negative = input.starts_with('-');
+    let binary_part = if is_negative {
+        &input[3..]
+    } else {
+        &input[2..]
+    };
+
+    if is_negative {
+        i64::from_str_radix(binary_part, 2).map(|n| Integer::from(-n))
+    } else {
+        u64::from_str_radix(binary_part, 2).map(Integer::from)
+    }
+}
```

Comment:
The only difference between these is the radix, so could be reduced into a single `parse_integer(lex, radix)`